### PR TITLE
fix: synchronize when resetting the keepalive timer

### LINKED ISSUE #20: runtime error: racy use of timers
Got this while running IPFS, but I figured it probably belongs here:

```
panic: runtime error: racy use of timers

goroutine 247502 [running]:
time.resetTimer(0xc003d3ff48, 0x3fc2d5cfdd34)
        /usr/local/go/src/runtime/time.go:233 +0x35
time.(*Timer).Reset(0xc003d3ff40, 0x6fc23ac00, 0x0)
        /usr/local/go/src/time/sleep.go:126 +0x7d
github.com/libp2p/go-yamux.(*Session).startKeepalive.func1()
        /home/leo/go/pkg/mod/github.com/libp2p/go-yamux@v1.3.1/session.go:345 +0x19e
created by time.goFunc
        /usr/local/go/src/time/sleep.go:168 +0x44
```

This is on go-ipfs commit 67741060c4332d27b6434559db3e6270444b1125, the daemon says this on startup:
```
go-ipfs version: 0.5.0-dev-67741060c
Repo version: 7
System version: amd64/linux
Golang version: go1.14
```

It's actually WSL, I built without CGO.

Here's another:
```
panic: runtime error: racy use of timers

goroutine 109005 [running]:
time.resetTimer(0xc002d589b8, 0x422c5ec82298)
        /usr/local/go/src/runtime/time.go:233 +0x35
time.(*Timer).Reset(0xc002d589b0, 0x6fc23ac00, 0xc002bd6300)
        /usr/local/go/src/time/sleep.go:126 +0x7d
github.com/libp2p/go-yamux.(*Session).recvLoop(0xc00393ef70, 0x0, 0x0)
        /home/leo/go/pkg/mod/github.com/libp2p/go-yamux@v1.3.1/session.go:516 +0x302
github.com/libp2p/go-yamux.(*Session).recv(0xc00393ef70)
        /home/leo/go/pkg/mod/github.com/libp2p/go-yamux@v1.3.1/session.go:483 +0x2b
created by github.com/libp2p/go-yamux.newSession
        /home/leo/go/pkg/mod/github.com/libp2p/go-yamux@v1.3.1/session.go:129 +0x379
```

### PR DESCRIPTION
**Pull request by Stebalien**
fixes #20

Also, fix the tests, test in CI 5 times and one more time with the race detector.

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -353,9 +358,18 @@ func (s *Session) stopKeepalive() {
...
 	if s.keepaliveTimer != nil {
 		s.keepaliveTimer.Stop()
+		s.keepaliveTimer = nil
 	}
 }
 
+func (s *Session) extendKeepalive() {
+	s.keepaliveLock.Lock()
+	if s.keepaliveTimer != nil && !s.keepaliveActive {
+		s.keepaliveTimer.Reset(s.config.KeepAliveInterval)
```
Comment (willscott) [inline session.go]: "Reset should be invoked only on stopped or expired timers with drained channels." - do we know that's the case here?

Comment (vyzo) [reply session.go]: already discussed in the previous pr...

File: session.go [hunk 2]
```diff
@@ -327,23 +328,27 @@ func (s *Session) startKeepalive() {
...
-
-		if s.keepaliveTimer == nil {
+		if s.keepaliveTimer == nil || s.keepaliveActive {
+			// keepalives have been stopped or a keepalive is active.
 			s.keepaliveLock.Unlock()
-			// keepalives have been stopped.
 			return
 		}
+		s.keepaliveActive = true
+		s.keepaliveLock.Unlock()
```
Comment (vyzo) [inline session.go]: this seems to be defered too? what's up with that?

File: session.go (hunk 1 above)
Comment (Stebalien) [reply session.go]: That's just bad documentation (https://github.com/libp2p/go-yamux/pull/16#discussion_r392064661). Reset will actually _call_ stop internally. The only reason you need to do it is if you need to drain the channel.

However, we're not even using a normal timer here.

File: session.go (hunk 2 above)
Comment (Stebalien) [reply session.go]: This code is in an anonymous function.

**Comment by Stebalien** [comment]
Ok, tests aren't perfect but they're _much_ better.

**Pull request merged**